[이진우] sprint10

ajantang opened the pull request:
## 요구사항

### 기본
#### 공통

- [x] Github에 위클리 미션 PR을 만들어 주세요.
- [x] React.js 혹은 Next.js를 사용해 진행합니다.
- [x] RESTful를 설계하고 백엔드 코드를 변경하세요.
  - [ ] (풀스택) 설계한 백엔드 코드에 맞게 프론트엔드 코드를 변경해 주세요.
https://panda-market-api.vercel.app의 API를 사용한 코드를 본인의 백엔드 API 코드로 변경하세요.
  - [x] (백엔드) 프론트엔드 코드에 맞게 백엔드 코드를 변경해 주세요.
- [ ] 백엔드 코드에 swagger를 추가해 API 명세 문서를 생성해 주세요.

#### 백엔드 구현 요구사항
##### 상품 등록

- [x] "상품 등록하기" 버튼 클릭 시, 상품 정보 등록 API 엔드포인트에 요청을 보내 상품을 등록합니다.
- [x] 상품 등록 시 필요한 필드(이름, 설명, 가격 등)의 유효성을 검증하는 미들웨어를 구현합니다.
- [x] multer 미들웨어를 사용하여 이미지 업로드 API를 구현해 주세요.
  - [x] 업로드된 이미지는 서버에 저장하고, 해당 이미지의 경로를 response 객체에 포함해 반환합니다.

##### 상품 상세

- [x] 상품을 조회할 때, 해당 상품에 대한 댓글 리스트, 사용자가 '좋아요'를 눌렀는지 여부를 확인할 수 있도록 응답 객체에 포함시켜 반환해 주세요.

##### 좋아요 기능

- [x] '좋아요' API를 만들어 주세요.
  - [x] 사용자는 상품 또는 게시글에 '좋아요'를 할 수 있습니다.
  - [x] $transaction을 사용해 주세요.
- [x] '좋아요' 취소 API를 만들어 주세요.
  - [x] 사용자는 상품 또는 게시글에 '좋아요'를 취소할 수 있습니다.
  - [x] $transaction을 사용해 주세요.
- [x] 상품 또는 게시글을 조회할 때, 사용자가 '좋아요'를 누른 항목인지 확인할 수 있도록 isLiked 필드를 응답 객체에 포함시켜 반환해 주세요.

##### 에러 처리

- [x] 모든 예외 상황을 처리할 수 있는 에러 핸들러 미들웨어를 구현합니다.
- [x] 서버 오류(500), 사용자 입력 오류(400 시리즈), 리소스 찾을 수 없음(404) 등 상황에 맞는 상태값을 반환합니다.

##### 라우트 중복 제거

- [x] 중복되는 라우트 경로(예: /users에 대한 get 및 post 요청)를 app.route()로 통합해 중복을 제거합니다.
- [x] express.Router()를 활용하여 중고마켓/자유게시판 관련 라우트를 별도의 모듈로 구분합니다.

##### 인증

- [x] User 스키마를 작성해 주세요.
  - [x] id, email, nickname, image, encryptedPassword, createdAt, updatedAt 필드를 가집니다.
- [x] 회원가입 API를 만들어 주세요.
  - [x] email, nickname, password 를 입력하여 회원가입을 진행합니다.
  - [x] password는 해싱해 저장합니다.
- [x] 로그인 API를 만들어 주세요.
  - [x] 사용자의 신원을 확인하고, 성공적인 인증 후에는 액세스 토큰을 발급해 response 객체에 포함해 반환합니다.

##### 상품 기능 인가

- [x] 로그인한 사용자만 상품을 등록할 수 있습니다.
- [x] 상품을 등록한 사용자만 해당 상품의 정보를 수정 및 삭제를 할 수 있습니다.
- [x] 로그인한 사용자만 상품에 '좋아요'를 추가하거나 삭제할 수 있습니다.

##### 게시글 기능 인가

- [x] 로그인한 사용자만 게시글을 등록할 수 있습니다.
- [x] 게시글을 등록한 사용자만 해당 게시글 정보를 수정 및 삭제할 수 있습니다.
- [x] 로그인한 사용자만 게시글에 '좋아요'를 추가하거나 삭제할 수 있습니다.

##### 댓글 기능 인가

- [ ] 로그인한 사용자만 상품에 댓글을 등록할 수 있습니다.
- [ ] 로그인한 사용자만 게시글에 댓글을 등록할 수 있습니다.
- [ ] 댓글을 등록한 사용자만 댓글을 수정하거나 삭제할 수 있습니다.

### 심화 요구사항


#### 인증

- [x] 토큰 기반 방식을 사용할 경우, 만료된 액세스 토큰을 새로 발급하는 리프레시 토큰 발급 기능을 구현합니다.(jwt sliding session 적용)

#### OAuth를 활용한 인증

- [ ] 구글 OAuth를 사용하여 회원가입 및 로그인 기능을 구현합니다.

#### 프로젝트 구조 변경

- [x] 프로젝트의 구조와 복잡성을 관리하기 위해 MVC 패턴이나 Layered Architecture와 같은 설계 방식을 적용해 보세요.

#### (생략 가능) 자유게시판 게시물 등록

- [x] 게시물 등록 시 필요한 필드(제목, 내용 등)의 유효성 검증하는 미들웨어를 구현합니다.
- [x] multer 미들웨어를 사용하여 이미지 업로드 API를 구현해 주세요.
  - [x] 업로드된 이미지는 서버에 저장하고, 해당 이미지의 경로를 response 객체에 포함해 반환합니다.

## 주요 변경사항

- then을 활용한 코드는 layered architecture 구조에서 함수의 비동기 여부를 판단하기 어려워 await으로 변경 중입니다.
- image는 google cloud에 연결했습니다. 다만, 연결에 인증에 필요한 json 파일이 필요해서 어떤 식으로 처리해서 git에 올릴지는 방법을 찾고 있습니다.(git에 노출되지 않도록 하고 싶습니다.)

## 스크린샷

![image](이미지url)

## 멘토에게

- 요구사항에 isLike로 좋아요 값을 요구했지만, 기존 프로젝트에서 사용하던 isFavorite으로 좋아요 값을 전달합니다

ajantang committed:
chore : 기본 스크립트 변경 postgreSQL 파일 실행

ajantang committed:
feat : add article's field favorite

미션 8에 요구사항에 맞춰 favorite 추가

ajantang committed:
feat : add article get option(keyword)

요구사항에 맞춰 키워트 기능 추가

ajantang committed:
style : change incorrect code

오타 수정

ajantang committed:
feat : add article's response(user name)

작성자 이름 추가

ajantang committed:
feat : apply select option in article API

select option 적용

ajantang committed:
feat : set default value of favorite

favorite 추가 후 api 로직 수정을 안해 생성 과정에서 favorite을 요구

ajantang committed:
feat : set default value of article

DB 스키마 수정 & 마이그레이션으로 해결이 안되서 row 생성과정에 default 값 추가

ajantang committed:
fix : query 기본 설정 오류

기본값을 설정하는 과정에서 입력된 값을 무시되는 현상 수정

ajantang committed:
feat : change reponse of article GET

name이 아닌 nickname으로 response 값 변경

ajantang committed:
feat : add reponse of article GET (myFavorite, user.image)

favorite과 profile image url 정보를 위한 res property 추가

ajantang committed:
test : add test code in http

BE 로직에 추가한게 아님

ajantang committed:
fix : incorrect mock data

unique 설정 data 중복 수정

ajantang committed:
feat : change response property of get article API

실수로 바꾼 get article response property 속성 복원
get article/:id response에 myFavorite 추가

ajantang committed:
feat : arrange schema

전반적인 DB 스키마 정리
mongoose code 삭제

ajantang committed:
feat : add auth & user api code

auth api code 추가 : 테스트 완료
user api code 추가 : GET /users/me/product, GET /users/me/favorite-products 테스트 필요(상품 데이터가 없어서 추후 테스트 예정)

ajantang committed:
feat : add product api code(unfinished)

GET /products 쿼리 테스트 필요. token 정보에 따라 favorite 정보 추가 필요요
GET /products/:id 테스트 필요. favorite 정보 추가 필요
PATCH /products/:id 테스트 필요

wildCodingWarrior commented on docs/DB.md:
```diff
@@ -0,0 +1,320 @@
+- max-length : 255
+- 참고 : 이메일의 국제 표준(Internet Email Format, RFC 5321)에 따르면, 전체 이메일 주소는 최대 254자까지 허용
+
+#### iamge
```
아마 오타겠죠?

wildCodingWarrior commented on src/middlewares/auth.js:
```diff
@@ -0,0 +1,171 @@
+              })
+            );
+          }
+
```
개인적으로는 try except를 사용한 async await 방식을 선호합니다만. 이 방법도 깔끔하네요.

wildCodingWarrior commented on src/routes/auth.js:
```diff
@@ -0,0 +1,75 @@
+   그냥 validateEmailPassword 유저 정보를 넘겨줌
+   넘겨주기는 하는네, 함수명을 생각하면 좋은건지는 모르겠네
+   */
+  const user = req.user;
```
이런 고민이 바로 Service, controller, repository 같은 레이어를 나누는 기준이 되겠습니다. 좋은 고민입니다.

wildCodingWarrior commented on src/routes/products.js:
```diff
@@ -0,0 +1,137 @@
+
+/** POST /products */
+productRouter.post("/", validateAccessToken, (req, res, next) => {
+  // images 선행 처리 필요. 따로 서버를 만들까...?
```
이미지만 처리하는 middleware를 만드는게 베스트겠지요.

wildCodingWarrior commented on src/routes/products.js:
```diff
@@ -0,0 +1,137 @@
+productRouter.patch("/:id", validateAccessToken, (req, res, next) => {
+  const { id } = req.params;
+  const { name, description, price, images, tags } = req.body;
+  const updateData = {
+    ...(name && { name }),
+    ...(description && { description }),
+    ...(price && { price }),
+    ...(images && { images }),
+    ...(tags && { tags }),
+  };
```
이 테크닉 아주 좋은 테크닉입니다. 해당하는 property가 존재할때만 넣는 방식이죠. 진우님은 이런 자바스크립트를 잘 활용하시는 모습이 많이 느껴집니다.

wildCodingWarrior reviewed (approved):
기능상의 구현은 사실 이미 잘 하고 계십니다만, 
멘토링 시간에 알려드렸던 레이어드 패턴 구현을 신경써보시는 것은 어떤가요? 지금도 Service와 repository, 단순 router로만 구분하더라도 코드가 훨씬 깔끔해질 것으로 생각이 됩니다.

ajantang committed:
feat : add image api

일단 요구사항에 맞춰 multer로 이미지 업로드

참고 : aws로 이미지 저장하도록 변경 예정

ajantang committed:
feat : change image res form & manage res form by seperate files

이미지 response url 배열로 url 정보 전달
db의 select 객체와 api res 형식을 responses, mappers 폴더의 js 파일을 이용해 관리

참고 : responses와 mappers로 무리하게 관리하는 부분은 삭제 예정

ajantang committed:
feat : add middleware to validate input value of product api

상품 등록/수정 api의 input 유효성 확인 middleware 추가

ajantang committed:
feat : add product, favorite product, product comment

상품, 상품 좋아요, 상품 댓글 api 추가

참고 : 상품 댓글 수정 테스트 필요

ajantang committed:
refactor : apply layered architecture on auth api

auth api에 layered architecture 적용

참고 기존 api도 적용 중

ajantang committed:
feat : upload image by using google cloud
>>
>> multer / google cloud를 이용한 이미지 업로드 로직으로 변경

ajantang committed:
feat : add post api code

layered architecture 구조에 맞춰 post(게시글) api 추가

참고 : 테스트 필요

wildCodingWarrior reviewed (approved):
아주아주 잘 하셨습니다. 코드의 응집도부터 레이어드 패턴 그 이상까지 아주 잘 적용하고 분할하셨네요. 일부 코드 구조가 중복되는 부분을 제외하면 좋습니다!
